bin/wasm-node/rust/src/json_rpc_service.rs

```diff
@@ -44,6 +44,65 @@ use std::{
     sync::{atomic, Arc},
 };
 
+/// Spawns a task to handle incoming JSON-RPC requests.
+pub async fn request_handling_task(
+    mut tasks_executor: Arc<Mutex<Box<dyn FnMut(Pin<Box<dyn Future<Output = ()> + Send>>) + Send>>>,
+    clients: Arc<HashMap<usize, Arc<JsonRpcService>>>,
+) {
+    (tasks_executor.clone().lock().await)(Box::pin(async move {
+        loop {
+            let ffi::Request {
+                json_rpc_request,
+                chain_index,
+            } = ffi::next_json_rpc().await;
+
+            // Each incoming request gets its own separate task.
+            let clients = clients.clone();
+            (tasks_executor.lock().await)(Box::pin(async move {
+                let request_string = match String::from_utf8(Vec::from(json_rpc_request)) {
+                    Ok(s) => s,
+                    Err(error) => {
+                        log::warn!(
+                            target: "json-rpc",
+                            "Failed to parse JSON-RPC query as UTF-8: {}", error
+                        );
+                        return;
+                    }
+                };
+
+                log::debug!(
+                    target: "json-rpc",
+                    "JSON-RPC => {:?}{}",
+                    if request_string.len() > 100 { &request_string[..100] } else { &request_string[..] },
+                    if request_string.len() > 100 { "…" } else { "" }
+                );
+
+                let (request_id, call) = match methods::parse_json_call(&request_string) {
+                    Ok(rq) => rq,
+                    Err(error) => {
+                        log::warn!(
+                            target: "json-rpc",
+                            "Ignoring malformed JSON-RPC call: {}", error
+                        );
+                        return;
+                    }
+                };
+
+                match clients.get(&chain_index).cloned() {
+                    Some(client) => client.handle_rpc(request_id, call).await,
+                    None => {
+                        log::warn!(
+                            target: "json-rpc",
+                            "Ignoring JSON-RPC call for chain index {}, for which a JSON-RPC service has not been started.",
+                            chain_index
+                        )
```

Comment:
This should send back a JSON-RPC error response actually. You can find how to do that elsewhere in the file.